fix decode_map throwing when `null` is being passed in

despairblue opened the pull request:
The current implementation of `decode_map` crashes when `null` is being passed in.

This can happen by using `json.decode`:

```gleam
import gleeunit
import gleeunit/should
import gleam/json
import gleam/dynamic

pub fn main() {
  gleeunit.main()
}

// gleeunit test functions end in `_test`
pub fn hello_world_test() {
  json.decode("null", dynamic.map(dynamic.string, dynamic.string))
  |> should.equal(Error(json.UnexpectedFormat([
    dynamic.DecodeError(expected: "Map", found: "null", path: []),
  ])))
}
```

This produces:
```
gleam_stdlib.mjs:636 Uncaught TypeError: Cannot convert undefined or null to object
    at Function.getPrototypeOf (<anonymous>)
    at decode_map (gleam_stdlib.mjs:636:24)
    at dynamic.mjs:599:7
```

I'm not sure how to write a test for this as:
* I can't install gleam/json, since it depends on stdlib, leading to a strange error during dependency resolution
* I can't create `null` with just the stdlib
* I haven't seen pure Erlang or JavaScript tests

Let me now what's the best way to test this.

inoas commented on src/gleam_stdlib.mjs:
```diff
@@ -564,6 +564,8 @@ export function classify_dynamic(data) {
   } else if (typeof data === "number") {
     return "Float";
+  } else if (data === null) {
+    return "null";
```
My gut tells me `Return Nil`: `https://github.com/gleam-lang/stdlib/blob/6cb76cb53d728ce32c9c1ab7424788e9aab65ca4/src/gleam_stdlib.mjs#L20`

despairblue replied:
I changed that. This was also the missing piece to add a test. Thanks 🙏🏾

lpil commented on src/gleam_stdlib.mjs:
```diff
@@ -564,6 +564,8 @@ export function classify_dynamic(data) {
   } else if (typeof data === "number") {
     return "Float";
+  } else if (data == null) {
+    return "Nil";
```
This should be `"null"` as `Nil` is `undefined` rather than `null`.

inoas replied:
See there I was just plain wrong. Bad gut!

despairblue replied:
We both were ^^

I'll change it to "Null" once I'm back at a computer.

I'll also change it to strict equality again, otherwise it would report that "Null" was passed in when "Nil" was passed in. 
I'll also add a check for "Nil" specifically.

Where I'm not sure is: Should I add a type called "Null" that maps to "null" or does that already exist?

despairblue replied:
Done, except for adding a new gleam type called `Null` for the JS type `null`.

lpil replied:
The string isn't the name of any specific type, rather it's more vaguely what you could call the value. There's no specific `Null` type in Gleam

despairblue replied:
> You've also added a case for Nil. That's cool, but could you add a test for the classify functin with that value too?

I've added deno tests for the `decode_map` function that cover that.

> We'll need to update the Erlang implementation also.

Done...I think. Not sure if you meant changing "Atom" to "Nil" or adding a case for "Nil" and adding "Atom". I don't know Erlang 😬

lpil replied:
`nil` specifically should be `Nil`, the rest should be `Atom`. it'll be something like this:

```erlang
classify(nil) -> <<"Nil">>;
classify(X) when is_atom(X) -> <<"Atom">>;
...
```

despairblue replied:
done abd0c9113c0b4f029fd86122d1885ce2b2590a01

lpil commented on src/gleam_stdlib.mjs:
```diff
@@ -633,6 +635,9 @@ export function decode_map(data) {
   if (data instanceof PMap) {
     return new Ok(PMap.fromMap(data));
   }
+  if (data == null) {
```
Could we also return this error if the typeof data is not object? It's still possible to crash this on older runtimes otherwise.

despairblue replied:
done

lpil reviewed (commented):
Thank you! I've left a couple small comments.

Could you also update the changelog with an `## Unreleased` section?

despairblue commented:
I rebased this on main.

despairblue commented:
@lpil I see that you added TS tests [here](https://github.com/gleam-lang/stdlib/commit/52d992c7e82a87b6a6d430ce565fad14ed7952c6).

I'll amend this to include TS tests for `null. For `undefined` we can keep the gleam tests. I'll also update the changelog to account for the new release and move my change to unreleased.

Until I got time to do that I'll move this to draft.

despairblue commented:
@lpil 

I've added deno tests for the `decode_map` function.
I changed the erlang implementation and also updated the changelog.

Oh I also changed the CI workflow to run the deno tests. I can't test it though without a maintainer a running the workflows.

Happy to pull the CI change out into it's own PR.

despairblue commented on src/gleam_stdlib.mjs:
```diff
@@ -392,11 +392,11 @@ export function regex_scan(regex, string) {
-        continue
+        submatches[n - 1] = new Some(match[n]);
+        continue;
       }
-      if(submatches.length > 0) {
-        submatches[n-1] = new None()
+      if (submatches.length > 0) {
+        submatches[n - 1] = new None();
```
Ooops. Didn't mean to commit these prettier changes. Should I leave them regardless?

lpil replied:
Fine to keep them :)

lpil commented on map/main_test.ts:
```diff
@@ -0,0 +1,24 @@
+      tail: {},
+    },
+  });
+});
```
Remove these please, testing should be done in Gleam.

despairblue replied:
The thing is I can't. Since there is no way to produce `null` I have to write the tests in JS/TS.

The implementation I had before was buggy, these tests were the only thing showing me that they were. 

I'm writing a very simple app and stumbled on multiple issues in the stdlib's JS implementation while doing this, I don't feel comfortable removing the tests here, because it could hurt gleam's early user experience if regressions are introduced.

How about I'm just removing the `undefined` test, because that one is covered by the gleam test I wrote. The `null` one I'm afraid has to stay or some parts of the JS code are not tested and a regression could be introduced at any time again.

Would that work?

despairblue replied:
I removed the redundant one and kept the one that can't be written in gleam: 393126e7cbc7a7a9591b581a5dfe5bec8aed602d

despairblue replied:
Alternatively I could also add an external function that returns `null` which would only be used for testing, e.g. not exported.

Would be better?

It would mean I'd need some help with Erlang though, as I'd also need a function in erlang that returns `atom`.

I assumed adding tests in TS would be OK, since there were added just recently for property testing.

lpil replied:
Yup, you got it. FFI can be used to construct `null` using a function in `gleam_stdlib_test_ffi.mjs`.

Creating an atom is a bit easier. Custom types without fields are atoms, so one can be made like so:

```gleam
pub type MyAtom {
  ThisIsAnAtom
}

pub fn atom_decode_test() {
  let atom = dynamic.from(ThisIsAnAtom)
  todo
}
```

despairblue replied:
Thanks for the example. I've implemented the test. I also remove the deno tests completely.

Let me know what you think :)

despairblue commented:
> Thank you! I've left a couple small comments.
> 
> Could you also update the changelog with an `## Unreleased` section?

That's already there: 

https://github.com/despairblue/stdlib/blob/393126e7cbc7a7a9591b581a5dfe5bec8aed602d/CHANGELOG.md?plain=1#L3-L5

Am I missing something?

I've also addressed the other things. Let me know what you think.

lpil commented:
Sorry about the slow reply! My inbox is super full at the moment!

despairblue commented on .github/workflows/ci.yml:
```diff
@@ -19,6 +19,10 @@ jobs:
+          deno-version: v1.x
       - run: gleam test --target erlang
       - run: gleam test --target javascript
+      - run: deno test
```
Shall I put them in another PR? Or you don't want to run the property tests in CI at all?

lpil replied:
Can just remove it, they were a debugging tool and likely will be removed soon.

despairblue replied:
I removed them from the PR.

lpil reviewed (approved):
Thank you!
